### PR TITLE
WIP - Redo CI and Include Automated Docker Deployment

### PR DESCRIPTION
**Pull request by VoR0220**
Signed-off-by: VoR0220 <rj@erisindustries.com>

This PR begins to address both #1572 and #1473 

Initial test: See if I can properly parallel OS X and reduce into each matrix each set of jobs. 
What I learned: No, I can't parallel OS X by putting each Xcode image into an array...I do need to separate them out clearer.

### CODE REVIEW

File: scripts/Dockerfile [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
 RUN \
-  apk --no-cache --update add build-base cmake boost-dev git                                                && \
-  sed -i -E -e 's/include <sys\/poll.h>/include <poll.h>/' /usr/include/boost/asio/detail/socket_types.hpp  && \
-  git clone --depth 1 --recursive -b develop https://github.com/ethereum/solidity                           && \
-  cd /solidity && cmake -DCMAKE_BUILD_TYPE=Release -DTESTS=0 -DSTATIC_LINKING=1                             && \
-  cd /solidity && make solc && install -s  solc/solc /usr/bin                                               && \
-  cd / && rm -rf solidity                                                                                   && \
-  apk del sed build-base git make cmake gcc g++ musl-dev curl-dev boost-dev                                 && \
+  ./scripts/install_deps.sh                                           									    										&& \
+  sed -i -E -e 's/include <sys\/poll.h>/include <poll.h>/' /usr/include/boost/asio/detail/socket_types.hpp  										&& \
```
Comment (BeOleg) [inline scripts/Dockerfile]: Really? sed the file to make it run correctly?

Comment (VoR0220) [reply scripts/Dockerfile]: Not to make it run correctly. Moreso to rid irritating warnings. Its been the case for some time and is the case with the current alpine docker file.

**Comment by chriseth** [comment]
@VoR0220 could you please explain what you are trying to do here? The matrix job as it stands now is way too heavy for travis. Also, I think we can restrict the testing to different compilers and different platforms (platforms, not distributions of the same platform).

**Comment by VoR0220** [comment]
Distributions i think could benefit here too. We hit more coverage this way. I was also thinking about what you said last night and my original plan didnt gel with what you had in mind here (my bad). However i still think that multiple versions and distros is a good idea due to Travis' limitations in these realms and to generally test whether or not our install scripts work. Currently my plan moving forward is to find a test filter to exclude all but the end to end tests. These self contained tests should run on all these distros and all these platforms and multiple versions on some of the major ones (os x, ubuntu). These tests are incredibly lightweight so its okay to have more. Finally, we run the end to end tests...which need modifications to just grab the generated bytecode when passed in a flag...nothing else. This gets rid of the containerization needs. The artifacts from this will then be gathered and (i haven't figured this part out yet) sent to an S3 instance or a shared repository of some kind where the artifacts are compared against the original one (assuming it passes). If the bytecode in both optimized and unoptimized format is equivalent, then the test passes, else it fails.

**Comment by chriseth** [comment]
@VoR0220 I don't think we should check that the install scripts work on every single commit. Again: This puts way to much strain on travis. We already have extremely high travis load because of the sheer amount of repositories in our organisation. Furthermore, we do not have the capacity to maintain that many distributions anyway. We can currently barely support ubuntu, emscripten and docker.

It makes total sense to check that we get exactly the same bytecode on windows, on linux and on emscripten, but that's about it.

**Comment by VoR0220** [comment]
@chriseth would it suffice to add alpine linux to that list since this is automating docker deployment?

File: scripts/build.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+if [ -z "$1" ]; then 
+    BUILD_TYPE=Release
+else 
+    BUILD_TYPE="$1"
+fi
+
+mkdir -p build
+cd build
+cmake .. -DCMAKE_BUILD_TYPE="$BUILD_TYPE"
+make -j2 && install -s solc/solc /usr/bin && install -s test/soltest
```
Comment (VoR0220) [inline scripts/build.sh:12]: remove -s

File: scripts/Dockerfile [hunk 3]
```diff
@@ -1,12 +1,21 @@
...
+#Determine whether or not we are in the testing suite
+ARG TEST=Off
+#Determine what build type we are using
+ARG BUILD=Release
+#Copy working directory on travis to the image
+COPY / $WORKDIR
+
+#Install dependencies, eliminate annoying warnings, test if we are in a test environment, if so, 
+#add in soltest to the binary. If not, assume this is basic solidity image, build it and ship it out.
+RUN ./scripts/install_deps.sh && sed -i -E -e 's/include <sys\/poll.h>/include <poll.h>/' /usr/include/boost/asio/detail/socket_types.hpp
```
Comment (VoR0220) [inline scripts/Dockerfile:16]: put into one run (makes image smaller)

File: scripts/run_ipc_tests.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+#mount volume so that host can access the tmp directory and the geth.ipc process
+docker run -di \
+	--volume /tmp:/tmp      \
+	--name testEth          \
+	ethereum/client-cpp     \
+	--test -d /tmp/testeth  
+
+echo "--> Running tests without optimizer..."
+soltest --run_test=SolidityOptimizer,GasMeterTests,SolidityEndToEndTest -- --ipcpath=/tmp/testeth/geth.ipc && \
```
Comment (VoR0220) [inline scripts/run_ipc_tests.sh:11]: change source for suites to add test label for "require IPC"

**Pull request closed without merging**